### ISSUE TITLE
Issues with SQLite Datastore

### ISSUE THREAD

**Issue by n3x77** [google/grr#497]
Hi @all,

after adding some more clients to grr (Version 3.1.0.2) we suddenly run into serious issues with grr and the sqlite data store. We increased the number of clients from 1200 to 1800. In the syslog we encounter problems that lead to the http_server process. Database files were locked and could not be written anymore. Therefore hunts get stuck and some databases went corrupt. 

We therefore tried to change the number of workers and http processes on the grr server. We tried different number of worker and http server combinations and started with 2 http_server processes that were load balanced by haproxy and 4 worker processes. But this did not help. So we added more http_server processes than workers because it seems that system limits of file handles were reached by the http_server processes. 

Sometime we see the following error in the log "May  8 10:30:42 grr_server[140154]:   File "/usr/lib/python2.7/SocketServer.py", line 290, in _handle_request_noblock
May  8 10:30:42 grr_server[140154]:     self.process_request(request, client_address)
May  8 10:30:42 grr_server[140154]:   File "/usr/lib/python2.7/SocketServer.py", line 607, in process_request
May  8 10:30:42 grr_server[140154]:     t.start()
May  8 10:30:42 grr_server[140154]:   File "/usr/lib/python2.7/threading.py", line 736, in start
May  8 10:30:42 grr_server[140154]:     _start_new_thread(self.__bootstrap, ())
May  8 10:30:42 6680 grr_server[140154]: error: can't start new thread
"

After checking the proc limits for the specific processes we can see that each http_process has reached the limit of 55000 open files. We changed some sysctl settings and also the /etc/limits.conf but this did not help.

On the hardware machine we can see that we do not have any performance issues regarding memory or disk space. Even the cpu load is not that high. 

Could you please give us some recommendations of which number of workers and http_servers we should use or the ratio between http_server processes and worker processes. It would be a great help for us to know if it is better to have more http_server processes or workers, or if there are more system limits we should adjust. 

When the error occurs we can see that me maximum number fs.file-max = 1048576 is not reached. We only have something about 100.000 file handles. 

Is there a change to stop all "active and hanging" hunts in grr via grr_console?

Thanks for your help

Cheers
Sven

**Comment by mbushkov** [comment]
Hi Sven!
Do I understand it correctly that all the processes run on a same machine? What was happening in GRR when the system got broken (i.e. did you run a hunt)?

The easiest way to stop all active and hanging hunts is to stop the hunts in GRR UI and then clean the notification queues in GRR console. Something along the lines of this snippet should clean the queues:
```
token = access_control.ACLToken(username="_")

from google3.ops.security.grr.lib import queue_manager
with queue_manager.QueueManager(token=token) as m:
  all_n = m.GetNotificationsForAllShards(rdfvalue.RDFURN("aff4:/H"))
  m.DeleteNotifications([n.session_id for n in all_n])
    
from google3.ops.security.grr.lib.hunts import results
to_delete = []
for subject, _ in data_store.DB.ScanAttributes(rdfvalue.RDFURN("aff4:/hunt_results_queue/Records"), ["aff4:sequential_value"], token=token):
  to_delete.append(subject)
results.HuntResultQueue.DeleteRecords(to_delete, token=token)
```

**Comment by n3x77** [comment]
Hi Misha,

thanks for your response. Yes we have a machine that has a lots of performance. Before the issues happened we had the standard configuartion that the installer script creates with only 1 http_server process and 2 worker processes. This worked well with 1100 clients connected to the server. We did normal hunting and ran cron hunts every 3 hours like e.g. looking for specific registry keys. The system had no performance issuses concerning load.

We then realized that after we connected 600 systems more that we got http errors with error code 500, and hunts stuck. In the log we could see that the requests from the clients were not written into the sqlite databases anymore and that write processes suddenly dropped. We decided to increase the number of http_server processes and the number of workers on the same machine. 

The question is is there a ratio you can suggest like 1 worker needs or 3 http_server processes or something like that? If we want to split the processes on various machines are we able to do this with the sqlite datastore or do we need to migrate to mysql?

Thanks for your help

**Comment by mbushkov** [comment]
Hi Sven,
I'd suggest migrating to MySQLAdvancedDataStore (https://github.com/google/grr/blob/e5bfaa59eae473a1aa65480e4f99f56400f60976/grr/lib/data_stores/mysql_advanced_data_store.py#L150). HTTP/SQLite data store is nice for quickly bootstrapping GRR, but it proved rather unstable under high load, leading to obscure issues and data corruption.

MySQLAdvancedDataStore should be much more stable and easier to maintain. Plus, you benefit from having the whole MySQL toolset at your disposal: i.e. you can easily do backups, monitor performance, etc.

Regarding the ratio: the issue you encountered seems to be caused by the datastore misbehaving. I.e. if workers/http_servers get stuck when accessing the datastore, it probably doesn't matter how many of worker/http_server processes are running, and increasing their number may even make matters worse. 

Nevertheless we normally run equal number of workers and http_servers. It may be a good idea also to ask other people on grr-users mailing list about the ratio their use, since the ratio may depend on a) the number of agents installed b) the type of the datastore used.

**Comment by n3x77** [comment]
Hi Misha,
thanks four your fast reply.

Is there a prefered way for migration from sqlite to mysql?

Cheers,
Sven

**Comment by mbushkov** [comment]
Hi Sven,
I'm afraid there's no established process of data migration - i.e. we have no scripts or tools for that. The easiest way is to switch to MySQLAdvancedDataStore with an empty MySQL database. All the active agents that you have installed will re-enroll within 10 minutes. This means that they will change their IDs, but will be functioning just as before.
If you need to preserve historical data, I'd suggest just backing up the SQLite datastore files.

When using MySQL datastore, the only criteria for splitting the processes between different machines is the performance characteristics of the hardware. If you can split between different machines, you effectively allow all the GRR processes to use more CPU/memory resources. At the same time, if you have a very powerful machine with lots of RAM and multiple cores, running everything on a same machine may work just fine. Then again, the requirements may change depending on the number of agents you have.

The only thing I'd suggest: given that all GRR processes are Python apps (meaning that they're effectively single-threaded, due to global interpreter lock), it makes sense to have roughly 1 CPU core per 1 GRR process (no matter if it's an admin ui, a worker or an http_server).

**Comment by n3x77** [comment]
Hi Misha,

Thanks,  I will try to use the advanced datastore. Checking the Commit you pasted, it seems that the Advanced Datastore Feature is not in the lastest GRR Relase of Version 3.1.0.2. What is your suggested way to get this Feature, installing grr from master branch?

Cheers,
Sven

**Comment by mbushkov** [comment]
Hi Sven,
The commit I referenced was the one from HEAD, but IIUC,
MySQLAdvancedDataStore was included into 3.1.0.2 release: see
https://github.com/google/grr/blob/server-3.1.0.2/grr/lib/data_stores/mysql_advanced_data_store.py

Can you please double check that you have it?

Cheers,
Misha

**Comment by n3x77** [comment]
Dear Misha,

thank you I will try it ! Thanks for your help! I think we can close the issue ;-)

Cheers
